express_webpack/index.html

```diff
@@ -99,50 +103,94 @@
         });
     });
 
+    /* E V E N T   L I S T E N E R S */
+    OneSignalDeferred.push(function(onesignal) {
+      onesignal.Notifications.addEventListener('permissionChange', function (isSubscribed) {
+        showEventAlert('permissionChange', { isSubscribed });
+      });
+    });
+
+    OneSignalDeferred.push(function(onesignal) {
+      onesignal.Notifications.addEventListener('permissionPromptDisplay', function (event) {
+        showEventAlert('permissionPromptDisplay', { event });
+      });
+    });
+
+    OneSignalDeferred.push(function(onesignal) {
+      onesignal.Notifications.addEventListener('dismiss', function (event) {
+        showEventAlert('dismiss', { event });
+      });
+    });
+
+    OneSignalDeferred.push(function(onesignal) {
+      onesignal.Notifications.addEventListener('willDisplay', function (event) {
+        showEventAlert('willDisplay', { event });
+      });
+    });
+
+    OneSignalDeferred.push(function(onesignal) {
+      onesignal.Notifications.addEventListener('click', function (event) {
+        showEventAlert('click', { event });
+      });
+    });
+
+    OneSignalDeferred.push(function(onesignal) {
+      onesignal.Slidedown.addEventListener('slidedownShown', function (event) {
+        showEventAlert('slidedownShown', { event });
+      });
+    });
+
+    // subscriptionChange
+    OneSignalDeferred.push(function(onesignal) {
+      onesignal.User.PushSubscription.addEventListener('subscriptionChange', function (event) {
+        showEventAlert('subscriptionChange', { event });
+      });
+    });
+
+    /* H E L P E R S */
+
+    function toggleShowEventAlert() {
+      showEventAlertToggleSetting = !showEventAlertToggleSetting;
+      document.getElementById('showEventAlertToggleSetting').innerText = showEventAlertToggleSetting ? 'true' : 'false';
+    }
+
+    function showEventAlert(eventName, payload) {
+      if (!showEventAlertToggleSetting) {
+        return;
+      }
+      alert(`OneSignal event ${eventName} fired!\n\nPayload:\n${JSON.stringify(payload, null, 2)}`);
+    }
+
     function sendUniqueOutcome() {
-        OneSignalDeferred.push(function(onesignal) {
-            const outcomeName = document.querySelector('#unique_outcome_name').value;
-            onesignal.sendUniqueOutcome(outcomeName);
-        });
+      const outcomeName = document.querySelector('#unique_outcome_name').value;
+      OneSignal.Session.sendUniqueOutcome(outcomeName);
```

Comment:
Possible race condition with OneSignal not being loaded yet. All OneSignal calls should go through `OneSignalDeferred.push`

Reply:
These are all triggered by button clicks so that's a non issue. Also, the problem with putting these in deferred, is the way you implemented it, there is no continuous processing of the functions pushed onto it, so they just get added to the array and nothing happens.

Maybe we need to revisit that.